futures-util/src/sink/unfold.rs

```diff
@@ -55,23 +61,24 @@ where
 
     fn start_send(self: Pin<&mut Self>, item: Item) -> Result<(), Self::Error> {
         let mut this = self.project();
-        debug_assert!(this.future.is_none());
-        let future = (this.function)(this.state.take().unwrap(), item);
-        this.future.set(Some(future));
+        let future = match this.state.as_mut().project_replace(State::Empty) {
+            StateProjOwn::Value(value) => (this.function)(value, item),
+            StateProjOwn::Future(_) | StateProjOwn::Empty => unreachable!()
```

Comment:
```suggestion
            StateProjOwn::Future(_) => panic!("poll_ready not called first"),
            StateProjOwn::Empty => unreachable!()
```

nit: `StateProjOwn::Empty` is really unreachable unless closure panics and the user catch it, but `StateProjOwn::Future` can be reached if the user makes a mistake. A different panic message may be useful.